### PR TITLE
Adds DeploymentConfig log functionality to Odo

### PR DESCRIPTION
**Pull request by cdrage**
This adds a new command `odo log` which will display the
DeploymentConfig logs for that particular component.

For example:

```sh
github.com/redhat-developer/odo  add-logs ✔                                                                                                                                                                                                                                                                                                                         15h55m
▶ ./odo list
ACTIVE     NAME        TYPE
*          nodejs3     nodejs

github.com/redhat-developer/odo  add-logs ✔                                                                                                                                                                                                                                                                                                                         15h57m
▶ ./odo log
Environment:
        DEV_MODE=false
        NODE_ENV=production
        DEBUG_PORT=5858
Launching via npm...
npm info it worked if it ends with ok
npm info using npm@3.10.9
npm info using node@v6.11.3
npm info lifecycle nodejs-ex@0.0.1~prestart: nodejs-ex@0.0.1
npm info lifecycle nodejs-ex@0.0.1~start: nodejs-ex@0.0.1

> nodejs-ex@0.0.1 start /opt/app-root/src
> node server.js

Server running on http://0.0.0.0:8080
```

### CODE REVIEW

**Comment by cdrage** [comment]
Quite simple (only retrieving DeploymentConfig logs).

But I believe it's a step in the right direction.

What do you think @kadel ?

```sh
github.com/redhat-developer/odo  add-logs ✔                                                                                                                                                                                                                                                                                                                         15h55m  
▶ ./odo list
ACTIVE     NAME        TYPE
*          nodejs3     nodejs

github.com/redhat-developer/odo  add-logs ✔                                                                                                                                                                                                                                                                                                                         15h57m  
▶ ./odo log
Environment: 
        DEV_MODE=false
        NODE_ENV=production
        DEBUG_PORT=5858
Launching via npm...
npm info it worked if it ends with ok
npm info using npm@3.10.9
npm info using node@v6.11.3
npm info lifecycle nodejs-ex@0.0.1~prestart: nodejs-ex@0.0.1
npm info lifecycle nodejs-ex@0.0.1~start: nodejs-ex@0.0.1

> nodejs-ex@0.0.1 start /opt/app-root/src
> node server.js

Server running on http://0.0.0.0:8080
```

File: pkg/occlient/occlient.go [hunk 1]
```diff
@@ -791,6 +795,33 @@ func (c *Client) FollowBuildLog(buildName string) error {
...
+	defer rd.Close()
+
+	// Create a new buffer so we can display the entire stream
+	buf := new(bytes.Buffer)
+
+	// Copy to buffer
+	_, err = io.Copy(buf, rd)
+
+	// Print
+	fmt.Println(buf.String())
```
[the remark this answers is not in the document]
Comment (cdrage) [reply pkg/occlient/occlient.go]: Okay, no worries. Should we change how to do https://github.com/redhat-developer/odo/blob/master/pkg/occlient/occlient.go#L1048 as well?

**Comment by cdrage** [comment]
Most changes corrected. Should we change https://github.com/redhat-developer/odo/blob/master/pkg/occlient/occlient.go#L1048 as well @kadel with regards to io.Writer? I'll implement it by passing in io.Writer from higher-up in the code.

File: pkg/occlient/occlient.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kadel) [reply pkg/occlient/occlient.go]: We should, but let's do that in another PR.

File: cmd/log.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+		// Application
+		currentApplication, err := application.GetCurrent(client)
+		checkError(err, "")
+
+		// Project
+		currentProject := project.GetCurrent(client)
+
+		// Retrieve and set the current component if none passed
+		var currentComponent string
+		if len(args) == 0 {
```
[the remark this answers is not in the document]
Comment (kadel) [reply cmd/log.go]: yes, but if `inputComponent` is empty than it returns the current component, otherwise it checks if inputComponent exists.

This is exactly what you are doing here with `args[0]` no?

wouldn't this have the same result?
```go
currentComponent := ""
if len(args) == 1 {
  currentComponent = args[1]
}
currentComponent = getComponent(client, currentComponent, currentApplication, currentProject)

 

```

File: pkg/occlient/occlient.go [hunk 3]
```diff
@@ -1020,7 +1021,7 @@ func (c *Client) WaitAndGetPod(selector string) (*corev1.Pod, error) {
 }
 
 // FollowBuildLog stream build log to stdout
-func (c *Client) FollowBuildLog(buildName string) error {
+func (c *Client) FollowBuildLog(name string) error {
```
Comment (kadel) [inline pkg/occlient/occlient.go]: why was this changed from `buildName` to just `name`?

File: pkg/occlient/occlient.go [hunk 4]
```diff
@@ -1046,8 +1047,35 @@ func (c *Client) FollowBuildLog(buildName string) error {
...
 	if _, err = io.Copy(stdout, rd); err != nil {
-		return errors.Wrapf(err, "error streaming logs for %s", buildName)
+		return errors.Wrapf(err, "error streaming logs for %s", name)
+	}
+
+	return nil
+}
+
+// Display DeploymentConfig log to stdout
+func (c *Client) DisplayDeploymentConfigLog(buildName string) error {
```
Comment (kadel) [inline pkg/occlient/occlient.go]: the argument is called `buildName`  but this function accepts DeploymentConfig name as an argument

**Comment by kadel** [comment]
> Most changes corrected. Should we change https://github.com/redhat-developer/odo/blob/master/pkg/occlient/occlient.go#L1048 as well @kadel with regards to io.Writer? I'll implement it by passing in io.Writer from higher-up in the code.

there already is an issue for that ;-) https://github.com/redhat-developer/odo/issues/312

File: cmd/log.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cdrage) [reply cmd/log.go]: @kadel We're passing in blank for currentComponent.. which feels as if it's a undocumented function since it will return the active component rather than "getting" the component that's being passed in (currentComponent). Does that make sense?

**Comment by cdrage** [comment]
@kadel Updated! I pass in an `io.Writer` now. I retrieve this using `color.Output` in order for us to get coloured output in Windows / macOS / Linux.

I have updated the function as well to write to io.Writer using `io.Copy`. This is all in yellow as well in order to differentiate and indicate to the user that this is raw OpenShift logs.

**Comment by cdrage** [comment]
Let me know what you think regarding `getComponent` and I'll change it to your suggestion.

File: cmd/log.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kadel) [reply cmd/log.go]: It is documented in a comment for that function.

I agree that it feels weird, but we have this logic in a lot of places, and it actually makes sense to have this abstracted away in one common function. Maybe a different name or small refactoring of that function might help.

File: pkg/component/component.go [hunk 5]
```diff
@@ -446,3 +446,17 @@ func GetComponentDesc(client *occlient.Client, currentComponent string, currentA
 
 	return componentType, path, componentURL, appStore, nil
 }
+
+// Get Component logs
+func GetLogs(client *occlient.Client, applicationName string) (string, error) {
```
[the remark this answers is not in the document]
Comment (cdrage) [reply pkg/component/component.go]: So for io.Writer, should we have this functionality higher-up, in the `cmd` package instead? At the moment, we have the "color" functionlity (using `color.Output`) within component.go rather than up in `cmd`. So perhaps I should move both.

[the remark this answers is not in the document]
Comment (kadel) [reply pkg/component/component.go]: Not sure if I understand that :-(

There are functions that will print output in two colors.
For example `component.Watch`.
It is called from `cmd` package and it takes `io.Writer` as argument.
It writes to `io.Writer` messages in white and also in yellow. So the color can't be set in `cmd` package.

**Comment by cdrage** [comment]
@kadel

I've done the following:

 - Added `os.Stdout` as a variable within `cmd/log.go` so we grab `io.Writer` at the most-highest abstraction level
 - Added `io.Writer` as parameters to both `GetLogs` as well as `DisplayDeploymentConfigLog`

File: cmd/log.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+			// Get the current compoment
+			currentComponent = getComponent(client, "", currentApplication, currentProject)
+
+		} else {
+
+			currentComponent = args[0]
```
Comment (kadel) [inline cmd/log.go]: we don't need this whole `else` section. This is already taken care of in `getComponent` function

**Comment by cdrage** [comment]
I've also gone ahead and added a small integration test.

File: cmd/log.go [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+		currentProject := project.GetCurrent(client)
+
+		// Retrieve and set the current component if none passed
+		var currentComponent string
+		if len(args) == 0 {
+
+			// Get the current compoment
+			currentComponent = getComponent(client, "", currentApplication, currentProject)
+
+		} else {
```
Comment (kadel) [inline cmd/log.go]: this whole else section is not needed. This logic is already in `getComponent`.

The whole point of using `getComponent` is that you can replace this entire `if else` with  following 5 lines 

```go
currentComponent := ""
if len(args) == 1 {
  currentComponent = args[0]
}
currentComponent = getComponent(client, currentComponent, currentApplication, currentProject)

``

File: pkg/component/component.go [hunk 8]
```diff
@@ -446,3 +446,17 @@ func GetComponentDesc(client *occlient.Client, currentComponent string, currentA
 
 	return componentType, path, componentURL, appStore, nil
 }
+
+// Get Component logs
+func GetLogs(client *occlient.Client, applicationName string, stdout io.Writer) (string, error) {
+
+	// Retrieve the stdout pipe to pass to occlient
```
Comment (kadel) [inline pkg/component/component.go]: is this comment still valid?

**Comment by cdrage** [comment]
@kadel Updated the comment :+1:

File: tests/e2e/e2e_test.go [hunk 9]
```diff
@@ -226,6 +226,12 @@ var _ = Describe("odoe2e", func() {
 				cmpSet := runCmd("odo component set nodejs")
 				Expect(cmpSet).To(ContainSubstring("nodejs"))
 			})
+
+			It("should be able to retrieve logs", func() {
+				runCmd("odo log")
+				runCmd("odo nodejs")
```
Comment (kadel) [inline tests/e2e/e2e_test.go]: `odo nodejs` what is this command?

File: tests/e2e/e2e_test.go [hunk 10]
```diff
@@ -226,6 +226,12 @@ var _ = Describe("odoe2e", func() {
 				cmpSet := runCmd("odo component set nodejs")
 				Expect(cmpSet).To(ContainSubstring("nodejs"))
 			})
+
+			It("should be able to retrieve logs", func() {
+				runCmd("odo log")
```
Comment (kadel) [inline tests/e2e/e2e_test.go]: this should probably test if something was returned

File: cmd/log.go [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+		// Application
+		currentApplication, err := application.GetCurrent(client)
+		checkError(err, "")
+
+		// Project
+		currentProject := project.GetCurrent(client)
+
+		// Retrieve and set the currentComponent
+		currentComponent := getComponent(client, args[0], currentApplication, currentProject)
```
Comment (kadel) [inline cmd/log.go]: It is failing here if no argument is passed to `odo log` command.

If there is no argument, you have to pass an empty string as `inputComponent` (second) parameter and `getComponent` will return the name of the current component. If `inputComponent`  parameter is non-empty string it will check that `inputComponent`  exists and return it back.

File: pkg/component/component.go [hunk 12]
```diff
@@ -446,3 +446,15 @@ func GetComponentDesc(client *occlient.Client, currentComponent string, currentA
 
 	return componentType, path, componentURL, appStore, nil
 }
+
+// Get Component logs
+func GetLogs(client *occlient.Client, applicationName string, stdout io.Writer) (string, error) {
```
Comment (kadel) [inline pkg/component/component.go]: We don't need to return `string`.

**Comment by kadel** [comment]
There is something wrong with output color, I see some lines in yellow some in white.

screen shot 2018-05-15 at 10 52 36

**Comment by kadel** [comment]
I found out what is happening with the color output.
Looks like wildfly is also using colored output (prints WARN messages in yellow) and it resets our settings.

**Comment by cdrage** [comment]
Updated!

Changes:

 - Fixed tests (tests check for exit codes anyways)
 - I found a nil error bug if there are *no* components that exist at all on OpenShift. I added this here: https://github.com/redhat-developer/odo/pull/429/files#diff-dace9c973da6801d7504aca65cf4da57R45 and within occlient here: https://github.com/redhat-developer/odo/pull/429/files#diff-6025e9b620e9321c03c11bf29e5c4dbdR1066)
 - Check to see if len(args) == 1 when picking a component (https://github.com/redhat-developer/odo/pull/429/files#diff-dace9c973da6801d7504aca65cf4da57R37)

File: pkg/occlient/occlient.go [hunk 13]
```diff
@@ -1052,6 +1053,43 @@ func (c *Client) FollowBuildLog(buildName string) error {
...
+
+	// RESTClient call to OpenShift
+	rd, err := c.appsClient.RESTClient().Get().
+		Namespace(c.namespace).
+		Name(deploymentConfigName).
+		Resource("deploymentconfigs").
+		SubResource("log").Stream()
+
+	if rd == nil {
+		return errors.New("Unable to retrieve DeploymentConfig from OpenShift, does your component exist?")
```
[the remark this answers is not in the document]
Comment (kadel) [reply pkg/occlient/occlient.go]: We already have nice user-facing message in `cmd`.

**Comment by cdrage** [comment]
@kadel got rid of the Uppercase! :+1:
